libqtile/widget/bluetooth.py

```diff
@@ -91,31 +77,30 @@ async def _init_device(self):
         introspect = await bus.introspect(BLUEZ, BLUEZ_PATH + self.hci)
         obj = bus.get_proxy_object(BLUEZ, BLUEZ_PATH + self.hci, introspect)
         iface = obj.get_interface(BLUEZ_DEVICE)
+        props = obj.get_interface(BLUEZ_PROPERTIES)
 
         connected = await iface.get_connected()
         name = await iface.get_name()
+        # subscribe receiver to property changed
+        props.on_properties_changed(self._signal_received_device)
         return connected, name
 
-    def _signal_received_adapter(self, message):
-        if message.message_type == MessageType.SIGNAL:
-            interface_name, changed_properties, invalidated_properties = message.body
-            powered = changed_properties.get('Powered', None)
-            if powered is not None:
-                self.powered = powered.value
-                self.update_text()
-
-    def _signal_received_device(self, message):
-        if message.message_type == MessageType.SIGNAL:
-            interface_name, changed_properties, invalidated_properties = message.body
-            connected = changed_properties.get('Connected', None)
-            if connected is not None:
-                self.connected = connected.value
-                self.update_text()
-
-            device = changed_properties.get('Name', None)
-            if device is not None:
-                self.device = device.value
-                self.update_text()
+    def _signal_received_adapter(self, interface_name, changed_properties, invalidated_properties):
```

Comment:
Looking good now.

Do you need two signal received methods? Given you're just looking for a particular key in the `changed_properties` it looks like you could handle in a single method.

Reply:
I believe they're both necessary, as the proxy objects are different.

Reply:
But you're not using the proxy object in the signal handler, you're just checking a dict for a certain key and retrieving the value. I don't see why different proxies can't send dicts to the same handler.

Reply:
Ah I see. The ```Name``` property is in both the device and the adapter, but adapter name doesn't ever change afaik, so should be fine.

Reply:
You can always check the `interface_name` as well if there's a clash with property names.